evm/p2p/test_peer.py

```diff
@@ -20,23 +19,109 @@
+# for them.
+HEADER_SYNC_TIMEOUT = 3
+
+
+class LESProtocolFull(LESProtocol):
+    _commands = [Status, Announce, BlockHeaders, BlockBodies, GetBlockHeaders, GetBlockBodies]
+
+    def send_announce(self, block_hash, block_number, total_difficulty, reorg_depth):
+        data = {
+            'head_hash': block_hash,
+            'head_number': block_number,
+            'head_td': total_difficulty,
+            'reorg_depth': reorg_depth,
+            'params': [],
+        }
+        header, body = Announce(self.cmd_id_offset).encode(data)
+        self.send(header, body)
+
+    def send_block_headers(self, headers, buffer_value, request_id):
+        data = {
+            'request_id': request_id,
+            'headers': headers,
+            'buffer_value': buffer_value,
+        }
+        header, body = BlockHeaders(self.cmd_id_offset).encode(data)
+        self.send(header, body)
+
+
+class LESPeerServing(LESPeer):
+    """A LESPeer that can send announcements and responds to GetBlockHeaders msgs.
+
+    Used to test our LESPeer implementation. Tests should call .send_announce(), optionally
+    specifying a block number to use as the chain's head and then use the helper function
+    wait_for_head() to wait until the client peer has synced all headers up to the announced head.
+    """
+    conn_idle_timeout = 2
+    reply_timeout = 1
+    max_headers_fetch = 20
+    _supported_sub_protocols = [LESProtocolFull]
+    _head_number = None
+
+    @property
+    def head_number(self):
+        if self._head_number is not None:
+            return self._head_number
+        else:
+            return self.chaindb.get_canonical_head().block_number
+
+    def send_announce(self, head_number=None, reorg_depth=0):
+        if head_number is not None:
+            self._head_number = head_number
+        header = self.chaindb.get_canonical_block_header_by_number(self.head_number)
+        total_difficulty = self.chaindb.get_score(header.hash)
+        self.les_proto.send_announce(
+            header.hash, header.block_number, total_difficulty, reorg_depth)
+
+    def handle_msg(self, cmd, msg):
+        if isinstance(cmd, GetBlockHeaders):
+            self.new_get_block_headers(msg)
+        else:
+            super(LESPeerServing, self).handle_msg(cmd, msg)
+
+    def new_get_block_headers(self, msg):
+        query = msg['query']
+        headers = []
+        if query.reverse:
+            start = max(0, query.block - query.max_headers)
+            # Shift our range() limits by 1 because we want to include the requested block number
+            # in the list of block numbers.
+            block_numbers = reversed(range(start + 1, query.block + 1))
+        else:
+            end = min(self.head_number + 1, query.block + query.max_headers)
+            block_numbers = range(query.block, end)
+        for i in block_numbers:
+            headers.append(self.chaindb.get_canonical_block_header_by_number(i))
```

Comment:
Thoughts on changing this to:

```python
headers = tuple(
    self.chaindb.get_canonical_block_header_by_number(1)
    for i in block_number
)
```

Removes mutation and maintains a readability.  Would also be fine as a `map` call too.